### PR TITLE
Reorg Headers, Add Basic Auth for Elasticsearch

### PR DESCRIPTION
**Pull request by distortedsignal**
**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

/area outputs

/area tests

**What this PR does / why we need it**:

Elasticsearch primarily supports Basic Authentication as the means of authenticating to the server. This PR should allow users to use HTTP Basic Authentication with Elasticsearch. There was an attempt to make it easy to apply Basic Authentication to other providers as well, though no other output providers are configured in this PR.

This PR also cleans up the Post method in client.go to force the outputs to set the headers in their own Post methods.

**Which issue(s) this PR fixes**:

N/A

**Special notes for your reviewer**:

Previously: I'm going to get to the tests eventually. Trust me.

Now: Maybe I got to the tests? Maybe?

### CODE REVIEW

**Comment by distortedsignal** [comment]
I don't know if this will give anyone any kind of confidence in the quality of this change, but I got this output on my machine:

```sh
➜  falcosidekick git:(add-es-basic-auth) date && make falcosidekick && date && make test && date && make test-coverage && date
Sat Jun 19 01:31:01 PDT 2021
go build -gcflags all=-trimpath=/src -asmflags all=-trimpath=/src -a -installsuffix cgo -o falcosidekick .
Sat Jun 19 01:31:27 PDT 2021
go vet ./...
go test -v -race ./...
?   	github.com/falcosecurity/falcosidekick	[no test files]
=== RUN   TestNewAlertmanagerPayloadO
--- PASS: TestNewAlertmanagerPayloadO (0.00s)
=== RUN   TestNewClient
2021/06/19 01:31:34 [ERROR] : test - Bad Endpoint
--- PASS: TestNewClient (0.00s)
=== RUN   TestPost
2021/06/19 01:31:34 [ERROR] :  - Header missing (400)
2021/06/19 01:31:34 [ERROR] :  - Authentication Error (401)
2021/06/19 01:31:34 [ERROR] :  - Access Denied (403)
2021/06/19 01:31:34 [ERROR] :  - Resource not found (404)
2021/06/19 01:31:34 [ERROR] :  - Bad Request (422)
2021/06/19 01:31:34 [ERROR] :  - Exceeding post rate limit (429)
2021/06/19 01:31:34 [ERROR] :  - Unexpected Response  (502)
2021/06/19 01:31:34 [INFO]  :  - Post OK (200)
--- PASS: TestPost (0.01s)
=== RUN   TestAddHeader
2021/06/19 01:31:34 [INFO]  :  - Post OK (200)
--- PASS: TestAddHeader (0.00s)
=== RUN   TestAddBasicAuth
2021/06/19 01:31:34 [INFO]  :  - Post OK (200)
--- PASS: TestAddBasicAuth (0.00s)
=== RUN   TestMutualTlsPost
2021/06/19 01:31:48 [INFO]  :  - Post OK (200)
--- PASS: TestMutualTlsPost (14.07s)
=== RUN   TestNewDatadogPayload
--- PASS: TestNewDatadogPayload (0.00s)
=== RUN   TestNewDiscordPayload
--- PASS: TestNewDiscordPayload (0.00s)
=== RUN   TestNewGoogleChatPayload
--- PASS: TestNewGoogleChatPayload (0.00s)
=== RUN   TestNewInfluxdbPayload
--- PASS: TestNewInfluxdbPayload (0.00s)
=== RUN   TestNewLokiPayload
--- PASS: TestNewLokiPayload (0.00s)
=== RUN   TestMattermostPayload
--- PASS: TestMattermostPayload (0.00s)
=== RUN   TestNewOpsgeniePayload
--- PASS: TestNewOpsgeniePayload (0.00s)
=== RUN   TestPagerdutyPayload
--- PASS: TestPagerdutyPayload (0.00s)
=== RUN   TestNewRocketchatPayload
--- PASS: TestNewRocketchatPayload (0.00s)
=== RUN   TestNewSlackPayload
--- PASS: TestNewSlackPayload (0.00s)
=== RUN   TestNewTeamsPayload
--- PASS: TestNewTeamsPayload (0.00s)
PASS
ok  	github.com/falcosecurity/falcosidekick/outputs	14.752s
=== RUN   TestPriorityType_MarshalJSON
=== RUN   TestPriorityType_MarshalJSON/Default
=== RUN   TestPriorityType_MarshalJSON/Debug
=== RUN   TestPriorityType_MarshalJSON/Informational
=== RUN   TestPriorityType_MarshalJSON/Notice
=== RUN   TestPriorityType_MarshalJSON/Warning
=== RUN   TestPriorityType_MarshalJSON/Error
=== RUN   TestPriorityType_MarshalJSON/Critical
=== RUN   TestPriorityType_MarshalJSON/Alert
=== RUN   TestPriorityType_MarshalJSON/Emergency
=== RUN   TestPriorityType_MarshalJSON/Unknown_Key
--- PASS: TestPriorityType_MarshalJSON (0.00s)
    --- PASS: TestPriorityType_MarshalJSON/Default (0.00s)
    --- PASS: TestPriorityType_MarshalJSON/Debug (0.00s)
    --- PASS: TestPriorityType_MarshalJSON/Informational (0.00s)
    --- PASS: TestPriorityType_MarshalJSON/Notice (0.00s)
    --- PASS: TestPriorityType_MarshalJSON/Warning (0.00s)
    --- PASS: TestPriorityType_MarshalJSON/Error (0.00s)
    --- PASS: TestPriorityType_MarshalJSON/Critical (0.00s)
    --- PASS: TestPriorityType_MarshalJSON/Alert (0.00s)
    --- PASS: TestPriorityType_MarshalJSON/Emergency (0.00s)
    --- PASS: TestPriorityType_MarshalJSON/Unknown_Key (0.00s)
=== RUN   TestPriorityType_UnmarshalJSON
=== RUN   TestPriorityType_UnmarshalJSON/Default
=== RUN   TestPriorityType_UnmarshalJSON/Debug
=== RUN   TestPriorityType_UnmarshalJSON/Informational
=== RUN   TestPriorityType_UnmarshalJSON/Notice
=== RUN   TestPriorityType_UnmarshalJSON/Warning
=== RUN   TestPriorityType_UnmarshalJSON/Error
=== RUN   TestPriorityType_UnmarshalJSON/Critical
=== RUN   TestPriorityType_UnmarshalJSON/Alert
=== RUN   TestPriorityType_UnmarshalJSON/Emergency
=== RUN   TestPriorityType_UnmarshalJSON/Unknown_Key
=== RUN   TestPriorityType_UnmarshalJSON/an_error
--- PASS: TestPriorityType_UnmarshalJSON (0.00s)
    --- PASS: TestPriorityType_UnmarshalJSON/Default (0.00s)
    --- PASS: TestPriorityType_UnmarshalJSON/Debug (0.00s)
    --- PASS: TestPriorityType_UnmarshalJSON/Informational (0.00s)
    --- PASS: TestPriorityType_UnmarshalJSON/Notice (0.00s)
    --- PASS: TestPriorityType_UnmarshalJSON/Warning (0.00s)
    --- PASS: TestPriorityType_UnmarshalJSON/Error (0.00s)
    --- PASS: TestPriorityType_UnmarshalJSON/Critical (0.00s)
    --- PASS: TestPriorityType_UnmarshalJSON/Alert (0.00s)
    --- PASS: TestPriorityType_UnmarshalJSON/Emergency (0.00s)
    --- PASS: TestPriorityType_UnmarshalJSON/Unknown_Key (0.00s)
    --- PASS: TestPriorityType_UnmarshalJSON/an_error (0.00s)
=== RUN   TestPriority
=== RUN   TestPriority/Default
=== RUN   TestPriority/Debug
=== RUN   TestPriority/Informational
=== RUN   TestPriority/Notice
=== RUN   TestPriority/Warning
=== RUN   TestPriority/Error
=== RUN   TestPriority/Critical
=== RUN   TestPriority/Alert
=== RUN   TestPriority/Emergency
=== RUN   TestPriority/Debug#01
=== RUN   TestPriority/Informational#01
=== RUN   TestPriority/Notice#01
=== RUN   TestPriority/Warning#01
=== RUN   TestPriority/Error#01
=== RUN   TestPriority/Critical#01
=== RUN   TestPriority/Alert#01
=== RUN   TestPriority/Emergency#01
=== RUN   TestPriority/Unknown_Key
--- PASS: TestPriority (0.01s)
    --- PASS: TestPriority/Default (0.00s)
    --- PASS: TestPriority/Debug (0.00s)
    --- PASS: TestPriority/Informational (0.00s)
    --- PASS: TestPriority/Notice (0.00s)
    --- PASS: TestPriority/Warning (0.00s)
    --- PASS: TestPriority/Error (0.00s)
    --- PASS: TestPriority/Critical (0.00s)
    --- PASS: TestPriority/Alert (0.00s)
    --- PASS: TestPriority/Emergency (0.00s)
    --- PASS: TestPriority/Debug#01 (0.00s)
    --- PASS: TestPriority/Informational#01 (0.00s)
    --- PASS: TestPriority/Notice#01 (0.00s)
    --- PASS: TestPriority/Warning#01 (0.00s)
    --- PASS: TestPriority/Error#01 (0.00s)
    --- PASS: TestPriority/Critical#01 (0.00s)
    --- PASS: TestPriority/Alert#01 (0.00s)
    --- PASS: TestPriority/Emergency#01 (0.00s)
    --- PASS: TestPriority/Unknown_Key (0.00s)
PASS
ok  	github.com/falcosecurity/falcosidekick/types	(cached)
Sat Jun 19 01:31:48 PDT 2021
go test ./outputs -count=1 -cover -v ./...
=== RUN   TestNewAlertmanagerPayloadO
--- PASS: TestNewAlertmanagerPayloadO (0.00s)
=== RUN   TestNewClient
2021/06/19 01:32:00 [ERROR] : test - Bad Endpoint
--- PASS: TestNewClient (0.00s)
=== RUN   TestPost
2021/06/19 01:32:00 [INFO]  :  - Post OK (200)
2021/06/19 01:32:00 [ERROR] :  - Header missing (400)
2021/06/19 01:32:00 [ERROR] :  - Authentication Error (401)
2021/06/19 01:32:00 [ERROR] :  - Access Denied (403)
2021/06/19 01:32:00 [ERROR] :  - Resource not found (404)
2021/06/19 01:32:00 [ERROR] :  - Bad Request (422)
2021/06/19 01:32:00 [ERROR] :  - Exceeding post rate limit (429)
2021/06/19 01:32:00 [ERROR] :  - Unexpected Response  (502)
--- PASS: TestPost (0.00s)
=== RUN   TestAddHeader
2021/06/19 01:32:00 [INFO]  :  - Post OK (200)
--- PASS: TestAddHeader (0.00s)
=== RUN   TestAddBasicAuth
2021/06/19 01:32:00 [INFO]  :  - Post OK (200)
--- PASS: TestAddBasicAuth (0.00s)
=== RUN   TestMutualTlsPost
2021/06/19 01:32:03 [INFO]  :  - Post OK (200)
--- PASS: TestMutualTlsPost (3.21s)
=== RUN   TestNewDatadogPayload
--- PASS: TestNewDatadogPayload (0.00s)
=== RUN   TestNewDiscordPayload
--- PASS: TestNewDiscordPayload (0.00s)
=== RUN   TestNewGoogleChatPayload
--- PASS: TestNewGoogleChatPayload (0.00s)
=== RUN   TestNewInfluxdbPayload
--- PASS: TestNewInfluxdbPayload (0.00s)
=== RUN   TestNewLokiPayload
--- PASS: TestNewLokiPayload (0.00s)
=== RUN   TestMattermostPayload
--- PASS: TestMattermostPayload (0.00s)
=== RUN   TestNewOpsgeniePayload
--- PASS: TestNewOpsgeniePayload (0.00s)
=== RUN   TestPagerdutyPayload
--- PASS: TestPagerdutyPayload (0.00s)
=== RUN   TestNewRocketchatPayload
--- PASS: TestNewRocketchatPayload (0.00s)
=== RUN   TestNewSlackPayload
--- PASS: TestNewSlackPayload (0.00s)
=== RUN   TestNewTeamsPayload
--- PASS: TestNewTeamsPayload (0.00s)
PASS
coverage: 25.9% of statements
ok  	github.com/falcosecurity/falcosidekick/outputs	3.600s	coverage: 25.9% of statements
Sat Jun 19 01:32:03 PDT 2021
```

File: outputs/gcpcloudrun.go [hunk 1]
```diff
@@ -10,6 +10,10 @@ import (
 func (c *Client) CloudRunFunctionPost(falcopayload types.FalcoPayload) {
 	c.Stats.GCPCloudRun.Add(Total, 1)
 
+	if c.Config.GCP.CloudRun.JWT != "" {
+		c.AddHeader(AuthorizationHeaderKey, "Bearer "+c.Config.GCP.CloudRun.JWT)
+	}
```
Comment (distortedsignal) [inline outputs/gcpcloudrun.go:15]: Compare to the following hunk:
https://github.com/falcosecurity/falcosidekick/blob/970a90504eb3d761c8cc9617a9153a9397ba5dbf/outputs/client.go#L181-L183

File: outputs/kubeless.go [hunk 2]
```diff
@@ -72,6 +80,11 @@ func (c *Client) KubelessCall(falcopayload types.FalcoPayload) {
 		}
 		log.Printf("[INFO]  : Kubeless - Function Response : %v\n", string(rawbody))
 	} else {
+		c.AddHeader(KubelessEventIDKey, uuid.New().String())
+		c.AddHeader(KubelessEventTypeKey, KubelessEventTypeValue)
+		c.AddHeader(KubelessEventNamespaceKey, c.Config.Kubeless.Namespace)
```
Comment (distortedsignal) [inline outputs/kubeless.go:85]: Compare to the following hunk in the current main branch:
https://github.com/falcosecurity/falcosidekick/blob/970a90504eb3d761c8cc9617a9153a9397ba5dbf/outputs/client.go#L175-L179

File: outputs/kubeless.go [hunk 3]
```diff
@@ -72,6 +80,11 @@ func (c *Client) KubelessCall(falcopayload types.FalcoPayload) {
...
+		c.ContentType = KubelessContentType
```
Comment (distortedsignal) [inline outputs/kubeless.go:86]: Compare to the following hunk in the current main branch:
https://github.com/falcosecurity/falcosidekick/blob/970a90504eb3d761c8cc9617a9153a9397ba5dbf/outputs/client.go#L165-L168

File: outputs/loki.go [hunk 4]
```diff
@@ -47,6 +50,7 @@ func newLokiPayload(falcopayload types.FalcoPayload, config *types.Configuration
 // LokiPost posts event to Loki
 func (c *Client) LokiPost(falcopayload types.FalcoPayload) {
 	c.Stats.Loki.Add(Total, 1)
+	c.ContentType = LokiContentType
```
Comment (distortedsignal) [inline outputs/loki.go:53]: Compare to the following hunk in the current main branch:
https://github.com/falcosecurity/falcosidekick/blob/970a90504eb3d761c8cc9617a9153a9397ba5dbf/outputs/client.go#L165-L168

File: outputs/opsgenie.go [hunk 5]
```diff
@@ -51,6 +51,7 @@ func newOpsgeniePayload(falcopayload types.FalcoPayload, config *types.Configura
 // OpsgeniePost posts event to OpsGenie
 func (c *Client) OpsgeniePost(falcopayload types.FalcoPayload) {
 	c.Stats.Opsgenie.Add(Total, 1)
+	c.AddHeader(AuthorizationHeaderKey, "GenieKey "+c.Config.Opsgenie.APIKey)
```
Comment (distortedsignal) [inline outputs/opsgenie.go:54]: Compare to the following hunk in the current main branch:
https://github.com/falcosecurity/falcosidekick/blob/970a90504eb3d761c8cc9617a9153a9397ba5dbf/outputs/client.go#L171-L173

File: outputs/webhook.go [hunk 6]
```diff
@@ -10,6 +10,12 @@ import (
 func (c *Client) WebhookPost(falcopayload types.FalcoPayload) {
 	c.Stats.Webhook.Add(Total, 1)
 
+	if len(c.Config.Webhook.CustomHeaders) != 0 {
+		for i, j := range c.Config.Webhook.CustomHeaders {
+			c.AddHeader(i, j)
+		}
+	}
```
Comment (distortedsignal) [inline outputs/webhook.go:17]: Compare to the following hunk in the current main branch:
https://github.com/falcosecurity/falcosidekick/blob/970a90504eb3d761c8cc9617a9153a9397ba5dbf/outputs/client.go#L187-L191

File: README.md [hunk 7]
```diff
@@ -487,6 +489,10 @@ care of lower/uppercases**) : `yaml: a.b --> envvar: A_B` :
   `false`)
 - **ELASTICSEARCH_CHECKCERT** : check if ssl certificate of the output is valid (default:
   `true`)
+- **ELASTICSEARCH_USERNAME** : use this username to authenticate to Elasticsearch if the
+  username is not empty (default: "")
+- **ELASTICSEARCH_PASSWORD** : use this password to authenticate to Elasticsearch if the
+  password is not empty (default: "")
```
Comment (Issif) [inline README.md:495]: can you sync descriptions between files please

**Review by Issif — CHANGES_REQUESTED** [review]
Really nice PR, thank you a lot. I had in my todo list for a while to refactore `client.go`. I was just too lazy for 😛 .

It's really nice to add sources for comparison, it's so convenient for the review, thank you.

I have a just a small comment for consistency of descriptions between files, else is good

File: README.md (hunk 7 above)
Comment (distortedsignal) [reply README.md:495]: Sure - which description would you prefer?

Comment (Issif) [reply README.md:495]: `use this (username|passworrd) to authenticate to Elasticsearch if not empty (default: "")`

seems good to me, wdyt?

Comment (distortedsignal) [reply README.md:495]: Ok, I think I got to this. Let me know if you're not happy with the wording. I can change it. :)

**Comment by distortedsignal** [comment]
The last commit message should be "Standardize _wording_ across documentation", but I can't spell right now.

**Comment by Issif** [comment]
> The last commit message should be "Standardize _wording_ across documentation", but I can't spell right now.

You can fix the last commit with `git commit --amend && git push origin add-es-basic-auth --force`

**Review by Issif — APPROVED** [review]
well done 👍

**Pull request merged**